### ISSUE TITLE
Deterministic hang when handling user events

### ISSUE THREAD

**Issue by hainest** [dyninst/dyninst#658]
Executing

    ./test_driver -64 -none -gcc -dynamiclink -create -pic -test pc_thread

works perfectly fine. But executing

    ./test_driver -64 -none -gcc -dynamiclink -create -nonpic -test pc_thread

causes a deterministic hang on a variety of platforms: Power8 with gcc7.3.1 and glibc-2.17, Power9 with gcc-7.3.1 and glibc-2.17, and x86 gcc-8.3 with glibc-2.23 (I haven't tried on ARM).

On x86, I see that [supportsUserThreadEvents](https://github.com/dyninst/testsuite/blob/master/src/proccontrol/pc_thread.C#L436) returns false. Later, test_driver then checks [isTerminated](https://github.com/dyninst/testsuite/blob/master/src/proccontrol/proccontrol_comp.C#L911)​​ which returns false and so it assumes the process must be blocking for a user event. However, the launched process is already terminated. The log says: "Process 11035 not terminated, is stopped, is not exited, blocking for events". It then goes on to try to block for events, but the process never responds and hangs on the call to [select](https://github.com/dyninst/testsuite/blob/master/src/proccontrol/proccontrol_comp.C#L1362).

At least one of the following must be going on:

1. proccontrol is not correctly determining if the process can handle user events
2. proccontrol is correct and the process cannot handle user events, and thus test_driver needs check this before blocking
    - But then how would the suspending/tracking process then be terminated?
3. proccontrol is not correctly determining that the process is terminated

Why this is happening with PIC and not with nonPIC, I have no idea.

Ping @sashanicolas and @mxz297.

**Comment by hainest** [comment]
Confirmed this affects ARM, as well. Debugging shows the same sequence to recreate the issue as on x86.

**Comment by mxz297** [comment]
For me, `./test_driver -64 -none -gcc -dynamiclink -create -nonpic -test pc_thread` passes on zeroah, but it deterministically hangs on power 9 (llnl.cs.rice.edu).

**Comment by hainest** [comment]
@mxz297 Is it hanging at the same spot?

**Comment by mxz297** [comment]
Will debug it in the weekend.

**Comment by hainest** [comment]
`./test_driver -64 -none -gcc -dynamiclink -create -pic -test test_thread_2` is also hanging on ARM and Power8.

**Comment by mxz297** [comment]
This turns out to be test suite problem. The test suite fails to continue or terminate the mutatee when it decides to skip the test. 

I posted a PR for this test suite issue (https://github.com/dyninst/testsuite/pull/107).

**Comment by hainest** [comment]
That PR fixes the `pc_thread` hangs. Now I'm seeing the `test_thread_2` hang on x86 in addition to ARM and PPC. Debugging shows that [`threadCreateCB`](https://github.com/dyninst/testsuite/blob/master/src/dyninst/test_thread_2.C#L125) is never invoked, so it gets an infinite loop [here](https://github.com/dyninst/testsuite/blob/master/src/dyninst/test_thread_2.C#L177).

Invoked with

```
export DYNINSTAPI_RT_LIB=/home/tim/paradyn/NoM_0kIQ/testsuite/tests/../dyninst/lib/libdyninstAPI_RT.so
export OMP_NUM_THREADS=2

LD_LIBRARY_PATH=/usr/local/lib/boost-1.69/lib:/home/tim/paradyn/libraries/lib:/home/tim/paradyn/NoM_0kIQ/dyninst/lib:/home/tim/paradyn/NoM_0kIQ/testsuite/tests ./test_driver -64 -none -gcc -dynamiclink -create -pic -test test_thread_2 -log tmp.log
```

gdb output shows `libthread_db` is found and loaded (this is on my home machine).

```
Starting program: /home/tim/paradyn/NoM_0kIQ/testsuite/tests/test_driver -64 -none -gcc -dynamiclink -create -pic -test test_thread_2 -log tmp.log
[Detaching after fork from child process 19503]
[Detaching after fork from child process 19506]
[Detaching after fork from child process 19509]
[Detaching after fork from child process 19512]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff0155700 (LWP 19524)]
[Detaching after fork from child process 19525]
[New Thread 0x7fffef74a700 (LWP 19527)]
Thread DB process constructor
[New Thread 0x7fffeef49700 (LWP 19528)]
[Detaching after fork from child process 19529]
[New Thread 0x7fffee748700 (LWP 19530)]
[Detaching after fork from child process 19531]
```

**Comment by mxz297** [comment]
@hainest I don't think your Dyninst installation finds `libpthread_db`. First of all, your gdb output only means that gdb finds `libpthread_db`; it does not mean Dyninst finds it. Then in your gdb output, there is the line `Thread DB process constructor`. See the code below:

https://github.com/dyninst/dyninst/blob/master/proccontrol/src/int_thread_db.C#L1801

The comment suggests that `Thread DB process constructor` will only be printed when `libpthread_db.so` is not found. Note that there is a `#else` above the code. So, I guess it is the Dyninst build system fails to find `libpthread_db.so` or fails to properly set macro. My guess the macro is about the code below:

https://github.com/dyninst/dyninst/blob/master/proccontrol/src/int_thread_db.C#L52

Can you check your Dyninst configuration/build log and see whether `cap_thread_db` is ever defined? And if `cap_thread_db` is not defined, why is it?

**Comment by mxz297** [comment]
I think the build system of finding `libpthread_db` has a bug. On llnl.cs.rice.edu (Power 9), there is no `cap_thread_db` defined, however there is `libpthread_db` on the system. Interestingly, in the CMakeCache file, there are lines:

```
//Details about finding Thread_Db
FIND_PACKAGE_MESSAGE_DETAILS_Thread_Db:INTERNAL=[/usr/lib64/libthread_db.so][/usr/include][v()]
```
So, I think the build system for setting macros has a bug.

**Comment by hainest** [comment]
@mxz297 In a fit of desperation, I reset the HEAD of Dyninst to the last known good state. Indeed, this fixed the hang issues. I moved the inclusion of `shared.cmake` above the global inclusion of `ThreadDB` to allow the compiler flags to be used when building the third-party dependencies from source. Unbeknownst to me, the `cmake/cap_arch_def.cmake` file conditionally sets `cap_thread_db` ([code](https://github.com/dyninst/dyninst/blob/master/cmake/cap_arch_def.cmake#L153)). This relies on a global ordering of inclusions which is a bug in itself. My change just exercised it.

This tells us three things:

1. `shared.cmake` is fundamentally broken in its design and needs to be fixed
2. ThreadDB is not an optional component since proccontrol doesn't work correctly when it is not present
3. Build system changes need to have the test suite run against them

On the upside, we fixed a couple of bugs in the test suite.

**Comment by mxz297** [comment]
> 2. ThreadDB is not an optional component since proccontrol doesn't work correctly when it is not present

ThreadDB is an optional component for proccontrol. Proccontrol contains process level and thread level control. ThreadDB is necessary for thread level control, but not process level. The actual bug is in the test suite, where the test suite does not properly skip thread level tests when ThreadDB is not found. 

> 3. Build system changes need to have the test suite run against them

The Dyninst build system change is orthogonal to the test suite issue.  The Dyninst build system should be able to find ThreadDB when it is available; I am not sure whether we need a test for this or not. The test suite should not hang when Dyninst believes there is not ThreadDB, even if Dyninst's belief is wrong.

> 
> On the upside, we fixed a couple of bugs in the test suite.

**Comment by hainest** [comment]
>ThreadDB is an optional component for proccontrol. Proccontrol contains process level and thread level control. ThreadDB is necessary for thread level control, but not process level.

Then proccontrol is broken because...

>The actual bug is in the test suite, where the test suite does not properly skip thread level tests when ThreadDB is not found.

There are still hangs in `test_thread_2`, `test_thread_3`, and `test_thread_6` when the test suite is fixed (https://github.com/dyninst/testsuite/pull/107), but ThreadDB is not compiled into Dyninst (i.e., #665 is not applied).

**Comment by mxz297** [comment]
> There are still hangs in `test_thread_2`, `test_thread_3`, and `test_thread_6` when the test suite is fixed ([dyninst/testsuite#107](https://github.com/dyninst/testsuite/pull/107)), but ThreadDB is not compiled into Dyninst (i.e., #665 is not applied).

`test_thread_*` series of tests are implemented in `testsuite/src/dyninst` while my posted fixes are implemented in `testsuite/src/proccontrol`. They are not the same code. 

Test `pc_thread` checks whether this test should execute or not, and will return SKIPPED if it should not. However, the proccontrol test infrastructure does not properly handle SKIPPED test.

I looked at the source code of `test_thread_2`; it does not ever bother to do the check. Still a test suite bug.

**Comment by hainest** [comment]
>I looked at the source code of test_thread_2; it does not ever bother to do the check. Still a test suite bug.

Ok. I'll file an issue over there. Should we just fold any additional fixes into dyninst/testsuite#107?

**Comment by mxz297** [comment]
> Ok. I'll file an issue over there. Should we just fold any additional fixes into [dyninst/testsuite#107](https://github.com/dyninst/testsuite/pull/107)?

OK. I will push fixes to the same branch.
